tests/simulators/carla/test_actions.py

```diff
@@ -0,0 +1,133 @@
+import os
+from pathlib import Path
+import signal
+import socket
+import subprocess
+import time
+
+import pytest
+
+try:
+    import carla
+
+    from scenic.simulators.carla import CarlaSimulator
+except ModuleNotFoundError:
+    pytest.skip("carla package not installed", allow_module_level=True)
+
+from tests.utils import compileScenic, sampleScene
+
+
+def checkCarlaPath():
+    CARLA_ROOT = os.environ.get("CARLA_ROOT")
+    if not CARLA_ROOT:
+        pytest.skip("CARLA_ROOT env variable not set.")
+    return CARLA_ROOT
+
+
+def launchCarlaServer():
+    CARLA_ROOT = checkCarlaPath()
+    subprocess.Popen(f"bash {CARLA_ROOT}/CarlaUE4.sh -RenderOffScreen", shell=True)
+
+    for _ in range(30):
+        if not isCarlaServerRunning():
+            time.sleep(1)
+
+    # extra 3 seconds to ensure server startup
+    time.sleep(3)
+    return
+
+
+def terminateCarlaServer():
+    result = subprocess.run("lsof -i :2000", capture_output=True, text=True, shell=True)
+    if result.stdout:
+        lines = result.stdout.strip().split("\n")
+        for line in lines[1:]:
+            parts = line.split()
+            pid = int(parts[1])
+            os.kill(pid, signal.SIGTERM)
+
+
+def isCarlaServerRunning(host="localhost", port=2000):
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+        sock.settimeout(1)  # Set a timeout of 1 second
+        try:
+            sock.connect((host, port))
+            return True
+        except (socket.timeout, ConnectionRefusedError):
+            return False
+
+
+@pytest.fixture(scope="package")
+def getCarlaSimulator(request):
+    if not isCarlaServerRunning():
+        launchCarlaServer()
+
+    def cleanup():
+        terminateCarlaServer()
+
+    request.addfinalizer(cleanup)
```

Comment:
Do you need to use an explicit finalizer here? I think it would work equally well to use a [yield fixture](https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#yield-fixtures-recommended), i.e., change the `return _getCarlaSimulator` to `yield _getCarlaSimulator` and then add the call to `terminateCarlaServer()` after that.

Reply:
Ack!